### PR TITLE
Envar Directives for Applications and Launchers

### PR DESCRIPTION
**Pull request by rhc54**
Signed-off-by: Ralph Castain <rhc@open-mpi.org>

### CODE REVIEW

**Comment by James-A-Clark** [comment]
Would it be worthwhile adding a 'set' LD_PRELOAD feature in addition to prepend and append? 

This may be necessary for working around issues where LD_PRELOAD is already set to something that is incompatible with a debugger for example. Or is there already a mechanism for setting arbitrary environment variables?

**Comment by jjhursey** [comment]
@mallen Might be interested in this discussion too. We have some similar functionality in Spectrum MPI that we rely on to manage the `LD_PRELOAD`, but in the ORTE PLM level. It has essentially a prepend and append like this PR proposes. If we can offload some of that logic to PMIx I think all the better.

**Comment by rhc54** [comment]
@James-A-Clark We don't have an attribute for setting/unsetting an envar as we assume someone provides those values when calling "spawn". However, this situation is indeed different as we are now looking at a tool wanting to adjust the environment specification prior to launch.

Your comment also sparked a thought in my otherwise dim mind: LD_PRELOAD is indeed just an envar. So perhaps we should generalize this RFC to define methods for a tool to set, delete, and modify envars as opposed to making it LD_PRELOAD-specific?

**Comment by markalle** [comment]
Thanks, I like it. As a wishlist I'd just ask for more of the same to cover even more use cases: I'd ask for the same thing for LD_LIBRARY_PATH and PATH, and I'd ask for an extra set of vars that do the same thing so multiple levels of users of this feature can segregate themselves from each other. 

That is, the main set of vars might be
PMIX_PREPEND_LD_PRELOAD
PMIX_APPEND_LD_PRELOAD
PMIX_PREPEND_LD_LIBRARY_PATH
PMIX_APPEND_LD_LIBRARY_PATH
PMIX_PREPEND_PATH
PMIX_APPEND_PATH
and on top of that a nice-to-have feature would be
PMIX_PREPEND_LD_PRELOAD_MPI
PMIX_APPEND_LD_PRELOAD_MPI
PMIX_PREPEND_LD_LIBRARY_PATH_MPI
PMIX_APPEND_LD_LIBRARY_PATH_MPI
PMIX_PREPEND_PATH_MPI
PMIX_APPEND_PATH_MPI
or maybe _SYSTEM or _DISTRO or some other suffix. The idea there is that maybe there are multiple users who want to use the feature and that makes it easier for them to all put their own item into the list without stepping on each other.

I don't see the code that implements the current PMIX_PREPEND/APPEND_LD_PRELOAD, but I want to also make sure everybody remembers the separation character differs for LD_PRELOAD (space) and PATH / LD_LIBRARY_PATH (colon).

**Comment by rhc54** [comment]
I think that this argues even more to just create something that lets you append, prepend, set, and delete an envar as opposed to creating something that is this specific. If people want to ensure they don't step on each other, than use prepend/append - these will always set the envar if it doesn't already exist, so no harm in using them.

I'm not sure I want to hard-code the details of how the various envars are concatenated. If you want to "append" to LD_PRELOAD, then maybe you should append " mylib" to ensure the space is present. Likewise, "append" LD_LIBRARY_PATH might be ":mypath". Have to think about it. I suppose we could add something to specify the separator - that might not be too horrible.

**Comment by James-A-Clark** [comment]
I agree that a generic feature to specify append, set, prepend and delete for any environment variable seems like the best solution now.

It also sounds simpler and cleaner to add the separator in the string that is passed to the API, rather than another function to set a custom separator.

**Comment by rhc54** [comment]
What if we step back for a minute and look at this from the use-case perspective? We obviously don't need anything for the case where someone calls PMIx_Spawn to start an application as they already have full control of the application's envars in that API. Likewise, we don't need an API for an app to change its own envars.

What we are really dealing with here are two cases.

**Case 1: a tool (e.g. a debugger) is being used to directly launch an application**
In this case, the tool is going to call PMIx_Spawn for the application itself (i.e., not using a launcher such as mpiexec in-between), thereby asking the RM's launcher to actually start the processes. PMIx_Spawn provides controls for specifying the application's environment, but not for modifying the launcher. Thus, we need to provide the debugger with a means to instruct the launcher in some way (e.g., replacing fork/exec with a "local spawn agent"), without impacting the application itself.

This use-case is probably best resolved by adding attributes (e.g., PMIX_LOCAL_SPAWN_AGENT) that the debugger can add to the job_info array in the PMIx_Spawn API. I see no need for a new API here.

**Case 2: a tool (e.g. a debugger) is being used to indirectly launch an application**
 In this case, PMIx_Spawn is given "mpirun" as the cmd to execute, and the tool has control over that environment. However, the app appears solely in the argv of mpirun and thus the debugger has no direct way of controlling the app's environment separate from the launcher's. I believe this is the original use-case raised by @James-A-Clark and friends as it correlates to a debugger starting mpirun and then needing to tell mpirun to change things (such as LD_PRELOAD) prior to launching its application.

Note that this second use-case also requires the ability to pass instructions to the launcher itself (e.g., replacing fork/exec) that are not covered by envars.

In this case, I think we can use the PMIx_Job_control API to allow the tool to pass control directives to mpirun. We again would only need to add attributes for this purpose, and can utilize the PMIx_Job_control_nb API to communicate those directives to mpirun once it is started (and prior to it launching the application).

This leaves the question of how to best specify an envar for the application itself, given that the tool has no direct control over the application's environment. The RFC already provides attributes for setting/unsetting/modifying envars, but lacks control over the separator. I agree with @James-A-Clark that we definitely don't want another function to set value separators for envars - that would be too cumbersome.

What I propose is that we define a new PMIx structure for specifying envar operations:
```
typedef struct {
    char *envar;
    char *value;
    char separator;
} pmix_envar_t;
```
and that we provide a set of attributes for controlling envars using that structure:

- PMIX_SET_ENVAR: sets the envar to the given value, overwriting any pre-existing one
- PMIX_UNSET_ENVAR: unset the envar, if present
- PMIX_PREPEND_ENVAR: prepend the given value to the specified envar using the separator character, creating the envar if it doesn't already exist
- PMIX_APPEND_ENVAR: append the given value as above

The resulting code might look like this:
```
pmix_envar_t foo = {
    "FOO_APPLES",
    "myvalue",
    ':'
};
pmix_info_t info;

PMIX_INFO_CONSTRUCT(&info);
PMIX_INFO_LOAD(&info, PMIX_PREPEND_ENVAR, &foo, PMIX_ENVAR);
PMIx_Job_control_nb(NULL, 0, &info, 1, cbfunc, cbdata);
```
The cases outlined in the RFC and discussion now become special cases of this more generalized capability. If this makes sense to people, I'll modify the RFC and prototype the code.

**Comment by James-A-Clark** [comment]
Looks good to me.

**Comment by rhc54** [comment]
All: I have formally submitted this for review, but we need a second from an organization other than Intel and Allinea/ARM.

File: RFC0022.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ * the aggregate value
+ */
+typedef struct {
+    char *envar;
+    char *value;
+    char separator;
+} pmix_envar_t;
+
+/* environmental variable operation attributes */
+#define PMIX_SET_ENVAR          "pmix.envar.set"          // (pmix_envar_t*) set the envar to the given value,
```
Comment (jjhursey) [inline RFC0022.md:94]: If we model this after `setenv` then we need a `PMIX_SET_ENVAR` (which overwrites if pre-exiting value exists) and `PMIX_SET_ENVAR_NO_OVERWRITE` (which does not).

Comment (jjhursey) [reply RFC0022.md:94]: Missed `PMIX_ADD_ENVAR` is the 'no overwrite' version. So nevermind that comment.

**Pull request merged**